### DISCUSSION TITLE
How are field validation (Pb admin ui) done and best practice for form validation ?

### DISCUSSION THREAD

**Discussion by madupuis90** [pocketbase/pocketbase#3040] [Q&A] [answered]
Hi, I love the project and I was going through the docs/code to figures this out but couldn't find my answer (no experience with Go)

I was wondering what happens under the hood whenever I set constraints on fields in the PB admin UI e.g. Text field min length. Does it do a check constraint at the DB level or does it implements validation at the service layer or something else ?

I was also wondering if there were best practices for form validation ? Initially I was thinking of going with a Sveltekit SSR and validate forms service side in Node, but after my last question I decided to go the SPA CSR route and I am not sure about the best practice for validation using Pocketbase. I was thinking a mix of field constraint + OnRecordBeforeCreateRequest hooks to validation cross field validation (e.g. if field 1 is set, make sure field 2 isn't empty)

Thanks

**Comment by ganigeorgiev** [comment, marked as answer]
> Does it do a check constraint at the DB level or does it implements validation at the service layer or something else ?

Most validators (email, url, non-empty, valid relation id, etc.) are on app level.
We rely on the DB errors for the unique index constraint checks.

> I was thinking a mix of field constraint + OnRecordBeforeCreateRequest hooks to validation cross field validation (e.g. if field 1 is set, make sure field 2 isn't empty)

This is a fine approach. Just keep in mind that the regular record crud operations are triggered also when making changes from the Admin UI because the same APIs are used (_to get the current auth state you can check https://pocketbase.io/docs/js-routing/#retrieving-the-current-auth-state; the http context object in `*Request` hooks is available under the `e.HttpContext` field_).

**Reply by madupuis90** [reply]
> Just keep in mind that the regular record crud operations are triggered also when making changes from the Admin UI because the same APIs are used 

Ah that's clever. This actually answer a question I didn't know I had
